File: robot-factory/typescript/tests/SupplierBuilder.ts
import { Money } from '../src/Money.js';
import type { PartOption } from '../src/PartOption.js';
import type { PartType } from '../src/PartType.js';
import { FakePartSupplier } from './FakePartSupplier.js';

interface PartEntry {
  type: PartType;
  option: PartOption;
  price: number;
}

export class SupplierBuilder {
  private _name = 'Supplier';
  private readonly parts: PartEntry[] = [];

  named(name: string): this { this._name = name; return this; }

  withPart(type: PartType, option: PartOption, price: number): this {
    this.parts.push({ type, option, price });
    return this;
  }

  build(): FakePartSupplier {
    const supplier = new FakePartSupplier(this._name);
    for (const { type, option, price } of this.parts) {
      supplier.withPart(type, option, new Money(price));
    }
    return supplier;
  }
}

File: robot-factory/python/tests/supplier_builder.py
from __future__ import annotations

from robot_factory import Money, PartOption, PartType

from .fake_part_supplier import FakePartSupplier


class SupplierBuilder:
    def __init__(self) -> None:
        self._name = "Supplier"
        self._parts: list[tuple[PartType, PartOption, float]] = []

    def named(self, name: str) -> SupplierBuilder:
        self._name = name
        return self

    def with_part(
        self, part_type: PartType, option: PartOption, price: float
    ) -> SupplierBuilder:
        self._parts.append((part_type, option, price))
        return self

    def build(self) -> FakePartSupplier:
        supplier = FakePartSupplier(self._name)
        for part_type, option, price in self._parts:
            supplier.with_part(part_type, option, Money(price))
        return supplier
